tensorflow/python/tools/optimize_for_inference_lib.py

```diff
@@ -531,3 +534,246 @@ def fuse_resize_and_conv(input_graph_def, output_node_names):
+                          {{"Const"},  //Variance
+                           {"Const"}   //Epsilon
+                          }
+                      }
+                   }, // end - Rsqrt
+                   {"Const"} //Gamma
+                  }
+                } //end - mul_1
+              }
+            } // end - mul_3
+          }
+        } //end - Sub
+      }
+  } //end - Add
+
+  Subgraph pattern when gamma value is 1 and the gamma scaling Mul is skipped
+  {"Add"
+      {{"Mul"           // mul_0
+          {{"*"},       // input to apply batch normalization
+           {"Rsqrt"     // same Rsqrt op used in Sub block
+              {"Add"
+                 {{"Const"},  //Variance
+                  {"Const"}   //Epsilon
+                 }
+              }
+            } // end - Rsqrt
+          }
+        }, // end - mul_0
+        {"Sub"
+          {{"Const"},    //Beta
+           {"Mul"        // mul_1
+              {{"Const"},  //Mean
+               {"Rsqrt"    //same Rsqrt op as in previous mul_0 block
+                  {"Add"
+                    {{"Const"},  //Variance
+                     {"Const"}   //Epsilon
+                    }
+                  }
+                } // end - Rsqrt
+              }
+           } // end - mul_1
+          }
+        } // end - Sub
+      }
+  } // end - Add
+
+  Args:
+    input_graph_def: A GraphDef containing a model.
+
+  Returns:
+    Modified graph with individual ops that made up of batch normalization
+    fused to FusedBatchNorm.
+
+  Raises:
+    ValueError: If the graph is badly formed with duplicate node names.
+  """
+  input_node_map = {}
+  for node in input_graph_def.node:
+    if node.name not in input_node_map:
+      input_node_map[node.name] = node
+    else:
+      raise ValueError("Duplicate node names detected for ", node.name)
+
+  # Check for data format and only proceed for NHWC or format not set case
+  data_format = None
+  for node in input_graph_def.node:
+    if "data_format" in node.attr.keys():
+      data_format = node.attr["data_format"]
+      break
+
+  if data_format is not None and data_format.s != b"NHWC":
+    tf_logging.warn("%s is not candidate for fusing decomposed batchnorm ops"
+                    % (data_format.s))
+    return input_graph_def
+
+  nodes_to_skip = {}
+  new_ops = []
+  for node in input_graph_def.node:
+    if node.op != "Add":
+      continue
+
+    # Add (Mul, Sub) or Add (Sub, Mul)
+    input0_op = node_from_map(input_node_map, node.input[0])
+    input1_op = node_from_map(input_node_map, node.input[1])
+
+    if input0_op.op == "Mul" and input1_op.op == "Sub":
+      data_scale_mul_op = input0_op
+      bias_mean_sub_op = input1_op
+    elif input0_op.op == "Sub" and input1_op.op == "Mul":
+      bias_mean_sub_op = input0_op
+      data_scale_mul_op = input1_op
+    else:
+      continue
+
+    # Mul (input, Mul)
+    input_data_op = node_from_map(input_node_map, data_scale_mul_op.input[0])
```

Comment:
I thought about this before, but choose not to add the communicative part as:
- the pattern is produced as a call to tf.nn.batch_normalization in earlier version, and the order of inputs is set.
- adding general communicative will require more checks in the code, and also, in this specific mul case, what if input_data_op is mul and it will add more complexity to trace upward for checking.

If the pattern won't have communicative property, it is preferred not to adding complexity to handle that use case. 
Is this ok with you?